File: src/main/java/eightyDays/java8/bank/account/Account.java
package eightyDays.java8.bank.account;

import eightyDays.java8.bank.partner.Identification;
import eightyDays.java8.bank.partner.Partner;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Currency;
import java.util.Date;
import java.util.List;

public abstract class Account {
    private Partner owner = null;
    private Identification number;
    private List<Booking> bookings = new ArrayList<>();

    Account(Partner owner) {
        this.number = Identification.newIdentification();
        this.owner = owner;
    }

    public Partner getOwner() {
        return owner;
    }

    public List<Booking> getBookings() {
        return bookings;
    }

    public Identification getNumber() {
        return number;
    }

    public BigDecimal getBalance() {
        return bookings.stream().map(Booking::getAmount).reduce(new BigDecimal(0), BigDecimal::add);
    }

    private void post(BigDecimal value, LocalDateTime valuta,  String text) {
        bookings.add(new Booking(value, valuta, text));
    }

    public void deposit(BigDecimal value) {
        deposit(value, LocalDateTime.now(), "Booking");
    }

    public void deposit(BigDecimal value, LocalDateTime valuta,  String text) {
        if (value.compareTo(BigDecimal.ZERO) > 0) {
            post(value, valuta, text);
        } else {
            throw new RuntimeException("Deposit of negative amount not allowed");
        }
    }

    public void withdraw(BigDecimal value) {
        withdraw(value, LocalDateTime.now(), "Booking");
    }

    public void withdraw(BigDecimal value, LocalDateTime valuta,  String text) {
        if (value.compareTo(BigDecimal.ZERO) < 0) {
            post(value, valuta, text);
        } else {
            throw new RuntimeException("Withdraw of negative or no amount not allowed");
        }
    }

    @Override
    public String toString() {
        return String.format("%s number:%s balance:%.2f", getClass().getSimpleName(), number.getNumber(), getBalance());
    }
}

File: steps/3/main/scala/eightyDays.scala/account/Account.scala
package eightyDays.scala.account

import eightyDays.scala.{Identification, Partner}
import scala.language.implicitConversions

abstract class Account(val owner: Partner,
                       val bookings: Seq[Booking],
                       factoryMethod: ((Partner, Seq[Booking]) => Account)) {

  val number = Identification()

  def balance: Amount = bookings.foldLeft(0: Amount)((balance, booking) => balance + booking)

  import java.time.LocalDateTime

  def withdraw(value: Amount, valuta: LocalDateTime = LocalDateTime.now()) =
    if (value > 0)
      post(-value, valuta)
    else
      throw new RuntimeException("Withdraw of negative amount not allowed")

  def deposit(value: Amount, valuta: LocalDateTime = LocalDateTime.now()) =
    if (value > 0)
      post(value, valuta)
    else
      throw new RuntimeException("Deposit of negative amount not allowed")

  protected[account] def post(value: Amount, valuta: LocalDateTime) = factoryMethod(owner, Booking(value, valuta) +: bookings)

  override def toString: String =
    s"${getClass.getSimpleName} number:${number.number} getBalance:$balance"
}